src/coreclr/jit/llvm.cpp

```diff
@@ -2435,11 +2468,34 @@ void Llvm::lowerToShadowStack()
 
                 CurrentRange().InsertBefore(node, retAddressLocal, storeNode);
             }
+
             if (node->OperIsLocalAddr())
             {
                 // Indicates that this local is to live on the LLVM frame, and will not participate in SSA.
                 _compiler->lvaGetDesc(node->AsLclVarCommon())->lvHasLocalAddr = 1;
             }
+            // check for implicit struct cast
+            else if (node->OperIsLocalStore())
```

Comment:
This should be `OperIs(GT_STORE_LCL_VAR)` for now. `STORE_LCL_FLD` is not supported for `TYP_STRUCT` currently.

That aside, I think this should be structured a little differently to avoid handling mismatched struct types in 
codegen -- convert the `LCL_VAR` source to `OBJ(LCL_VAR_ADDR)`, for `IND` sources, convert them to `OBJ<dest struct>` (same with mismatched `OBJ/BLK`).

Reply:
Thanks. 
I've removed the codegen conversion for `LCL_VAR` source.  I don't seem to have any failures after that.  For the `IND` sources problem, is there a test that you can think of that would exercise it?

Reply:
> is there a test that you can think of that would exercise it

Something like this should hit it:
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
private static StructWithIndex Problem(StructWithStructWithIndex* p, StructWithIndex b)
{
    b = p->StructWithIndex;
    JitUse(&b);

    return b;
}

struct StructWithIndex
{
    public int Index;
    public int Value;
}

struct StructWithStructWithIndex
{
    public StructWithIndex StructWithIndex;
    public int AnotherIndex;
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static void JitUse<T>(T* arg) where T : unmanaged { }
```
Basically, the Jit likes to dissolve `TYP_STRUCT` nodes into layout-less `IND TYP_STRUCT` nodes that only the parent nodes will know the true type of. This is most prevalent for the RHSs of struct stores (aka `OperIsBlockOp`).

Reply:
Thanks.  that looks good to me.  Before:
```
N001 (  3,  2) [000002] ------------         t2 =    LCL_VAR   int    V00 arg0         u:1 (last use) Zero Fseq[StructWithIndex]
                                                  /--*  t2     int
N002 (  6,  4) [000003] D--XG-------         t3 = *  IND       struct
                                                  /--*  t3     struct
N004 ( 10,  7) [000006] DA-XG-------              *  STORE_LCL_VAR struct<Program+StructWithIndex, 8>(AX) V01 arg1
```
After with latest change:
```
N001 (  3,  2) [000002] ------------         t2 =    LCL_VAR   int    V00 arg0         u:1 (last use) Zero Fseq[StructWithIndex]
                                                  /--*  t2     int
N002 (  6,  4) [000003] D--XG-------         t3 = *  OBJ       struct<Program+StructWithIndex, 8>
                                                  /--*  t3     struct
N004 ( 10,  7) [000006] DA-XG-------              *  STORE_LCL_VAR struct<Program+StructWithIndex, 8>(AX) V01 arg1
```

We can't handle return types of struct, so it fails the actual codegen at the moment, but the test is in at least.